manta-crypto/src/key/agreement.rs

```diff
@@ -0,0 +1,282 @@
+where
+    K: GenerateSecret<COM>,
+{
+    #[inline]
+    fn generate_secret(
+        &self,
+        public_key: &Self::PublicKey,
+        ephemeral_secret_key: &Self::EphemeralSecretKey,
+        compiler: &mut COM,
+    ) -> Self::SharedSecret {
+        (*self).generate_secret(public_key, ephemeral_secret_key, compiler)
+    }
+}
+
+/// Key Agreement
+pub trait Agree<COM = ()>: PublicKeyType + SecretKeyType + SharedSecretType {
+    /// Performs the agreement protocol on `public_key` and `secret_key` to arrive at the
+    /// [`SharedSecret`](SharedSecretType::SharedSecret).
+    fn agree(
+        &self,
+        public_key: &Self::PublicKey,
+        secret_key: &Self::SecretKey,
+        compiler: &mut COM,
+    ) -> Self::SharedSecret;
+}
+
+impl<K, COM> Agree<COM> for &K
+where
+    K: Agree<COM>,
+{
+    #[inline]
+    fn agree(
+        &self,
+        public_key: &Self::PublicKey,
+        secret_key: &Self::SecretKey,
+        compiler: &mut COM,
+    ) -> Self::SharedSecret {
+        (*self).agree(public_key, secret_key, compiler)
+    }
+}
+
+/// Key Agreement Secret Reconstruction
+pub trait ReconstructSecret<COM = ()>:
+    EphemeralPublicKeyType + SecretKeyType + SharedSecretType
+{
+    /// Performs the agreement protocol on `ephemeral_public_key` and `secret_key` to arrive at the
+    /// [`SharedSecret`](SharedSecretType::SharedSecret).
+    fn reconstruct_secret(
+        &self,
+        ephemeral_public_key: &Self::EphemeralPublicKey,
+        secret_key: &Self::SecretKey,
+        compiler: &mut COM,
+    ) -> Self::SharedSecret;
+}
+
+impl<K, COM> ReconstructSecret<COM> for &K
+where
+    K: ReconstructSecret<COM>,
+{
+    #[inline]
+    fn reconstruct_secret(
+        &self,
+        ephemeral_public_key: &Self::EphemeralPublicKey,
+        secret_key: &Self::SecretKey,
+        compiler: &mut COM,
+    ) -> Self::SharedSecret {
+        (*self).reconstruct_secret(ephemeral_public_key, secret_key, compiler)
+    }
+}
+
+/// Testing Framework
+#[cfg(feature = "test")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "test")))]
+pub mod test {
+    use super::*;
+    use crate::{
+        constraint::Satisfied,
+        eclair::{
+            bool::{Assert, AssertEq},
+            cmp::PartialEq,
+        },
+    };
+
+    /// Tests if the `agreement` property is satisfied for `K`.
+    #[inline]
+    pub fn agreement<K, COM>(scheme: &K, lhs: &K::SecretKey, rhs: &K::SecretKey, compiler: &mut COM)
+    where
+        COM: Assert + Satisfied,
+        K: Agree<COM> + Derive<COM>,
+        K::SharedSecret: PartialEq<K::SharedSecret, COM>,
+    {
+        let fst = scheme.agree(&scheme.derive(rhs, compiler), lhs, compiler);
+        let snd = scheme.agree(&scheme.derive(lhs, compiler), rhs, compiler);
+        compiler.assert_eq(&fst, &snd);
+        /*
+        assert!(
+            compiler.is_satisfied(),
+            "Compiler does not have all its constraints satisfied."
+        );
+        */
```

Comment:
Just delete these for now and remove `Satisfied` constraint, this pattern is not well established yet, I don't want to merge it until we figure out a good API.

Reply:
(same in the other test)